server/src/test/java/org/elasticsearch/cluster/health/ClusterHealthAllocationTests.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.cluster.health;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ESAllocationTestCase;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.cluster.node.DiscoveryNode;
+import org.elasticsearch.cluster.node.DiscoveryNodes;
+import org.elasticsearch.cluster.routing.RoutingTable;
+
+import java.util.Collections;
+
+import static org.elasticsearch.cluster.routing.ShardRoutingState.*;
+
+public class ClusterHealthAllocationTests extends ESAllocationTestCase {
+    public void testClusterHealth() {
+        ClusterState clusterState = ClusterState.builder(ClusterName.DEFAULT).build();
+        clusterState = addNode(clusterState, "node_m", true);
+        assertEquals(ClusterHealthStatus.GREEN, getClusterHealthStatus(clusterState));
+        MetaData metaData = MetaData.builder()
+            .put(IndexMetaData.builder("test")
+                .settings(settings(Version.CURRENT))
+                .numberOfShards(2)
+                .numberOfReplicas(1))
+            .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(metaData.index("test")).build();
+        clusterState = ClusterState.builder(clusterState).metaData(metaData).routingTable(routingTable).build();
+        MockAllocationService allocation = createAllocationService();
+        clusterState = allocation.reroute(clusterState, "reroute");
+        assertTrue(clusterState.nodes().getDataNodes().size() == 0);
+        assertTrue(clusterState.nodes().getMasterNodes().size() == 1);
+        assertEquals(ClusterHealthStatus.RED, getClusterHealthStatus(clusterState));
+        clusterState = addNode(clusterState, "node_d1", false);
+        assertTrue(clusterState.nodes().getDataNodes().size() == 1);
+        clusterState = allocation.reroute(clusterState, "reroute");
```

Comment:
As above, let's use `applyStartedShardsUntilNoChange` here instead of separate reroute/applyStartedShards steps.

Reply:
Hey @DaveCTurner  I think this wont work because we need to first change shards state to `INITIALIZING` state and then `applyStartedShards`. Using only `applyStartedShardsUntilNoChange` without `reroute` wont change the shard state to `INITIALIZING`.

Reply:
I'm surprised by this, because `applyStartedShardsUntilNoChange` repeatedly calls `startShardsAndReroute` which calls `applyStartedShards` and then `reroute`. If there are no `INITIALIZING` shards then the first `applyStartedShards` will do nothing, but the `reroute` should make some `INITIALIZING` shards that the next `applyStartedShards` will then move to `STARTED`.

You say you _think_ this won't work. Did you try it? If so, and it didn't work, then can you share your attempt so I can investigate further?

Reply:
Sorry @DaveCTurner I was using a previous version which didnt  `startInitializingShardsAndReroute`
have https://github.com/elastic/elasticsearch/blob/master/test/framework/src/main/java/org/elasticsearch/cluster/ESAllocationTestCase.java#L150

Let me pull it out with master . Thank you